drivers/gpio/gpio_sam.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2018 Justin Watson
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <gpio.h>
+
+#include "gpio_utils.h"
+
+typedef void (*config_func_t)(struct device *dev);
+
+struct gpio_sam_config {
+	Pio *regs;
+	config_func_t config_func;
+	u32_t periph_id;
+};
+
+struct gpio_sam_runtime {
+	sys_slist_t cb;
+};
+
+#define DEV_CFG(dev) \
+	((const struct gpio_sam_config *const)(dev)->config->config_info)
+
+static int gpio_sam_config(struct device *dev, int access_op, u32_t pin,
+			   int flags)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Pio *const pio = cfg->regs;
+	u32_t mask = 1 << pin;
+
+	if (access_op != GPIO_ACCESS_BY_PIN) {
+		return -ENOTSUP;
+	}
+
+	/* Setup the pin direcion. */
+	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_OUT) {
+		pio->PIO_OER = mask;
+	} else {
+		pio->PIO_ODR = mask;
+	}
+
+	/* Setup interrupt configuration. */
+	if (flags & GPIO_INT) {
+		/* Enable the interrupt. */
+		pio->PIO_IER = mask;
+
+		/* Enable the additional interrupt modes. */
+		pio->PIO_AIMER = mask;
+
+		switch (flags) {
+		case GPIO_INT_LEVEL:
+			pio->PIO_LSR = mask;
+			break;
+		case GPIO_INT_EDGE:
+			pio->PIO_ESR = mask;
+			break;
+		case GPIO_INT_DOUBLE_EDGE:
+			return -ENOTSUP;
+		default:
+			return -ENOTSUP;
+		}
+
+		if (flags & GPIO_INT_ACTIVE_HIGH) {
+			/* Set to high-level or rising edge. */
+			pio->PIO_REHLSR = mask;
+		} else {
+			/* Set to low-level or falling edge. */
+			pio->PIO_FELLSR = mask;
+		}
+	} else {
+		/* Disable the interrupt. */
+		pio->PIO_IDR = mask;
+
+		/* Disable ther additional interrupt modes. */
+		pio->PIO_AIMDR = mask;
+	}
+
+	/* Setup Pull-up resistor. */
+	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+		/* Enable pull-up. */
+		pio->PIO_PUER = mask;
+	} else {
+		pio->PIO_PUDR = mask;
+	}
+
+	/* Setup Pull-down resistor. */
+	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+		pio->PIO_PPDER = mask;
+	} else {
+		pio->PIO_PPDDR = mask;
+	}
+
+#ifdef SOC_SERIES_SAM3X
```

Comment:
nit: but since you have an #elif, this usually from a style point of view is #if defined(SOC_SERIES_SAM3X)